text/0000-pi-types.md

```diff
@@ -0,0 +1,379 @@
+         O = for<r..> TraitId<P1..Pn>      // Object type fragment
+         r = 'x                            // Region name
+    +    E = F(E)                          // Constant function application.
+    +      | p                             // const type parameter
+    +      | [...]                         // etc.
+
+Note that the `const` syntax is only used when declaring the parameter.
+
+## `impl` unification
+
+Only one `where` bound can be specified on each disjoint implementations (for
+possible extensions, see below).
+
+To find the right implementation, we use the data from the type inference (see
+the inference rules above). Since the parameters are, in fact, not much
+semantically different from normal generic parameters, we can resolve it is a
+normal manner.
+
+Likewise are disjointness checks based on structural equality.
+
+Since not all parameters' edges are necessarily the identity function,
+dispatching these would be undecidable. A way to solve this problem is to
+introduce some syntax allowing to specify the `impl` parameters.
+
+## An example
+
+This is the proposed syntax:
+
+```rust
+use std::{mem, ptr};
+
+// We start by declaring a struct which is value dependent.
+struct Array<const n: usize, T> {
+    // `n` is a constexpr, sharing similar behavior with `const`s, thus this
+    // is possible.
+    content: [T; n],
+}
+
+// We are interested in exploring the `where` clauses and Π-constructors:
+impl<const n: usize, T> Array<n, T> {
+    // This is simple statically checked indexing.
+    fn checked_index<const i: usize>(&self) -> &T where i < n {
+        //                 note that this is constexpr  ^^^^^
+        unsafe { self.content.unchecked_index(i) }
+    }
+
+    // "Push" a new element, incrementing its length **statically**.
+    fn push(self, elem: T) -> Array<n + 1, T> {
+        let mut new: [T; n + 1] = mem::uninitialized();
+        //               ^^^^^ constexpr
+        unsafe {
+            ptr::copy(self.content.as_ptr(), new.as_mut_ptr(), n);
+            ptr::write(new.as_mut_ptr().offset(n), elem);
+        }
+
+        // Don't call destructors.
+        mem::forget(self.content);
+
+        // So, the compiler knows the type of `new`. Thus, it can easily check
+        // if the return type is matching. By siply evaluation `n + 1`, then
+        // comparing against the given return type.
+        Array { content: new }
+    }
+}
+
+fn main() {
+    let array: Array<2, u32> = Array { content: [1, 2] };
+
+    assert_eq!(array.checked_index::<0>(), 1);
+    assert_eq!(array.checked_index::<1>(), 2);
+    assert_eq!(array.push(3).checked_index::<2>(), 3);
+}
+```
+
+# How we teach this
+
+This RFC aims to keep a "symmetric" syntax to the current construct, giving an
+intuitive behavior, however there are multiple things that are worth explaining
+and/or clearing up:
+
+1. What are Π-types? How are they different from those from other languages?
```

Comment:
I hope we don't need to be a type theorist before learning how to use this feature, like we didn't talk about linear logic or region based memory management when we introduce ownership/lifetime in The Book (Π? wha?).

I think the tutorial should start with intuitive use cases like `impl<T: Clone, const n: usize> Clone for [T; n]` and `fn concat<T, const m: usize, const n: usize>(left: [T; m], right: [T; n]) -> [T; m+n]`, and your example above.